crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2056,26 +2058,54 @@ impl<'db> ClassType<'db> {
     /// Returns the inferred type of the class member named `name`.
     ///
     /// Returns [`Symbol::Unbound`] if `name` cannot be found in this class's scope
-    /// directly. Use [`ClassType::class_member`] if you require a method that will
+    /// directly. Use [`Class::class_member`] if you require a method that will
     /// traverse through the MRO until it finds the member.
     pub(crate) fn own_class_member(self, db: &'db dyn Db, name: &str) -> Symbol<'db> {
         let scope = self.body_scope(db);
         symbol(db, scope, name)
     }
 }
 
+/// A singleton type representing a single class object at runtime.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct ClassLiteralType<'db> {
+    class: Class<'db>,
+}
+
+impl<'db> ClassLiteralType<'db> {
+    fn member(self, db: &'db dyn Db, name: &str) -> Symbol<'db> {
+        self.class.class_member(db, name)
+    }
+}
+
+impl<'db> From<ClassLiteralType<'db>> for Type<'db> {
+    fn from(value: ClassLiteralType<'db>) -> Self {
+        Self::ClassLiteral(value)
+    }
+}
+
+/// A type representing the set of runtime objects which are instances of a certain class.
+///
+/// Some specific instances are important enough that they really constitute their own type
+/// which should compare and hash differently to other instances of the same class.
```

Comment:
nit: I don't think "compare and hash" are really the behaviors that matter most here? I would write this in terms of type semantics.

```suggestion
/// Some specific instances of some types need to be treated specially by the type system; for example, various special forms are instances of `typing._SpecialForm`, but need to be handled differently in annotations.
```

Reply:
Yes, thank you! It was thinking through the "oh wait... two `InstanceType`s with the same class now compare and hash differently if one is a known instance and the other is not?" question that made me realise how uncomfortable I was with the new idiom. But that's a facet of our implementation rather than an aspect of what the type _represents_, of course.